### PR TITLE
Add GH action for running tests against upstream dev

### LINKED ISSUE #4574: nightly upstream test
As discussed in the call today it would be nice to remove the upstream test from our ci and setup a "nightly" job to test against the development versions of pandas etc. This should lead to less failures unrelated to specific PRs. Question: how are we going to see that the upstream build failed? Having the failure in random PRs is not clean but very visible...

This could be done via github actions. See xpublish for an example setup: https://github.com/xarray-contrib/xpublish/tree/master/.github/workflows

xref #4313

### PR DESCRIPTION
**Pull request by andersy005**
This GitHub action will run tests against the upstream dependencies. When the tests fail, the GH action will create an issue (with "CI" label). The body of the created issue includes:

- test summary report
- a link to the failed workflow run 

Here's a couple screenshots from a toy repository:

Screen Shot 2020-11-13 at 4 12 03 PM

Screen Shot 2020-11-13 at 4 12 13 PM

 - [x] Closes #4574
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
 - [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Comment by keewis** [comment]
thanks, @andersy005, this looks great.

For reference, this is how we currently install the upstream-dev dependencies:
https://github.com/pydata/xarray/blob/dd9fe2a8a414ddefa3b04b934163c9ccc628c5c7/ci/azure/install.yml#L13-L53

File: .github/workflows/upstream-dev-ci.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+name: CI
+on:
+  push:
+    branches:
+      - master
+  schedule:
+    - cron: '0 0 * * *' # Daily “At 00:00” GMT
```
Comment (max-sixty) [inline .github/workflows/upstream-dev-ci.yaml:7]:
```suggestion
    - cron: '0 0 * * *' # Daily “At 00:00” UTC
```
:)

**Comment by max-sixty** [comment]
This is awesome! Thanks for kicking it off @andersy005

**Comment by andersy005** [comment]
Of course, @keewis & @max-sixty! 


The latest commits include updates to the issue body and adds a build matrix for Python 3.7+. The issue body will look like:

Screen Shot 2020-11-14 at 10 16 13 PM

> For reference, this is how we currently install the upstream-dev dependencies:

Thank you for the pointer, @keewis! I added PyPI nightly wheels for scipy, numpy, pandas... What should do about the matploblib wheels issue?

File: ci/requirements/upstream-dev.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  - toolz
+  - pip:
+    - numbagg
+    - git+https://github.com/dask/dask 
+    - git+https://github.com/dask/distributed 
+    - git+https://github.com/zarr-developers/zarr 
+    - git+https://github.com/Unidata/cftime 
+    - git+https://github.com/mapbox/rasterio 
+    - git+https://github.com/hgrecco/pint 
+    - git+https://github.com/pydata/bottleneck
```
Comment (andersy005) [inline ci/requirements/upstream-dev.yml:49]: @keewis, let me know if these lines need to be replaced with

```bash
 python -m pip install \
        --no-deps \
        --upgrade \
        git+https://github.com/dask/dask \
        git+https://github.com/dask/distributed \
        git+https://github.com/zarr-developers/zarr \
        git+https://github.com/Unidata/cftime \
        git+https://github.com/mapbox/rasterio \
        git+https://github.com/hgrecco/pint \
        git+https://github.com/pydata/bottleneck
```
to match what is done in the existing upstream dev CI.

Comment (keewis) [reply ci/requirements/upstream-dev.yml:49]: Not sure. The current setup installs everything and then force-removes the ones we reinstall. That way, binary dependencies are installed using `conda` and only the libraries themselves are installed using `pip`, which might be advantageous for dependencies that don't have wheels.

If we decide to reproduce the current setup it would be nice to put the commands in a separate script and call that from both setups. In addition to keeping both CI in sync this would also help when trying to reproduce failures locally: right now setting up the environment involves copy/pasting commands.

**Comment by keewis** [comment]
> What should do about the matploblib wheels issue?

since it's broken right now I'd keep the version from `conda-forge` until the nightly wheels are available again

File: .github/workflows/upstream-dev-ci.yaml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  upstream-dev:
+    name: upstream-dev
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        shell: bash -l {0}
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.7", "3.8", "3.9"]
```
Comment (keewis) [inline .github/workflows/upstream-dev-ci.yaml:23]: the test matrix we have right now is `["3.6", "3.7", "3.8"]`. It is time to drop `3.6` and to start testing `3.9`, but maybe we should do that in a separate PR?

Comment (mathause) [reply .github/workflows/upstream-dev-ci.yaml:23]: I think we currently only do a `py38-upstream-dev` test. Not sure we actually need a matrix here.

File: .github/workflows/upstream-dev-ci.yaml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            const title = "⚠️ Nightly upstream-dev CI failed ⚠️"
+            const workflow_url = `https://github.com/${process.env.GITHUB_REPOSITORY}/actions/runs/${process.env.GITHUB_RUN_ID}`
+            const issue_body = `[Workflow Run URL](${workflow_url})\n${logs}`
+            github.issues.create({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: issue_body,
+              title: title,
+              labels: ['CI']
+             })
```
[the remark this answers is not in the document]
Comment (max-sixty) [reply .github/workflows/upstream-dev-ci.yaml:98]: I thought the same. One fairly manual resolution to this is to merge something xfail the test, and then a PR to un-xfail the test.

Otherwise it would have to search the existing issues — maybe possible though probably difficult?

[the remark this answers is not in the document]
Comment (andersy005) [reply .github/workflows/upstream-dev-ci.yaml:98]: > I might be missing something (especially since I don't know much about github script and action artifacts), but this looks like it will open a new issue on every run while the fix is being worked on. Is there a way to avoid that?



An issue will be created only when the following condition is met

```yaml
if: "always()&&(github.event_name == 'schedule')&&(github.repository == 'pydata/xarray')"
```

i.e., the workflow run has to be triggered via a `cron schedule`, and this trigger should be active on the `pydata/xarray` repository (which allows us to avoid opening issues on other people's folks). Currently, this condition has some limitations though. For instance, a new issue will be created every day as long as the upstream-dev CI is not passing.

> Otherwise it would have to search the existing issues — maybe possible though probably difficult?

I'm investigating whether there's a straightforward way to keep track of the latest issue created via the GitHub action, and adding comments to an existing issue instead of creating a new issue.

File: .github/workflows/upstream-dev-ci.yaml (hunk 3 above)
Comment (andersy005) [reply .github/workflows/upstream-dev-ci.yaml:23]: > I think we currently only do a py38-upstream-dev test. Not sure we actually need a matrix here.

If there is no need to test against multiple Python versions, I'm happy to remove the matrix, and stick with Python 3.8 only

File: .github/workflows/upstream-dev-ci.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (keewis) [reply .github/workflows/upstream-dev-ci.yaml:98]: > For instance, a new issue will be created every day as long as the upstream-dev CI is not passing.

Exactly, that's what I meant. Sorry for not being clear enough.

> I'm investigating whether there's a straightforward way to keep track of the latest issue created via the GitHub action, and adding comments to an existing issue instead of creating a new issue.

That would be cool, but I would like to make sure we don't post a comment every day, even if the failures don't change. Luckily, `pytest` prints the `short test summary info` section, so we should be able to extract the test ids and do a set comparison with the last comment (or issue, should we decide we need to open a new issue if different tests fail).

Actually, you could upload a file containing the extracted test ids as a artifact and use that to check whether the failed tests changed in the next run. Depending on how we compare, we might even be able to avoid opening a new issue for the remaining failures after merging a partial fix.

[the remark this answers is not in the document]
Comment (alexamici) [reply .github/workflows/upstream-dev-ci.yaml:98]: Personally I would skip opening an issue automatically in the initial PR. I would prefer people to monitor the workflow and report issue in a non automatic way.

If we see that this builds get overlooked, then we add the automatic reporting.

File: ci/requirements/upstream-dev.yml (hunk 2 above)
Comment (andersy005) [reply ci/requirements/upstream-dev.yml]: Addressed in [`dc6ba90` (#4583)](https://github.com/pydata/xarray/pull/4583/commits/dc6ba9083bab51f69e0da674f2db3f5b7a18bc01)

File: .github/workflows/upstream-dev-ci.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (max-sixty) [reply .github/workflows/upstream-dev-ci.yaml:97]: Do we think people are going to proactively check the overnight runs? Great if so, and if there's a realistic possibility, fine to try it.

File: .github/workflows/upstream-dev-ci.yaml [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                    labels: [variables.label]
+                })
+            } else {
+                github.issues.update({
+                    owner: variables.owner,
+                    repo: variables.name,
+                    issue_number: issue_info.number,
+                    body: issue_body
+                })
+            }
```
Comment (andersy005) [inline .github/workflows/upstream-dev-ci.yaml:129]: I am confident this gives us what we want in terms of reporting failures:, i.e.:

1. Find the most recently created issue (used for reporting failures)
2. If there's an existing and _open_ issue, update the body of the issue
3. Else if there's no candidate, _open_ issue, create a new issue and post the summary information of the failing tests

Comment (max-sixty) [reply .github/workflows/upstream-dev-ci.yaml:129]: Very impressive. I think this is the best of all worlds.

**Review by keewis — APPROVED** [review]
thanks, @andersy005, this looks good to me.

Updating the issue body is a nice idea to avoid getting notifications. I would have implemented a different behavior to make sure different failures are reported as different issues (as described in https://github.com/pydata/xarray/pull/4583#discussion_r524815461), but I think we can just try out how much of an issue the merging of all upstream-dev failures into a single issue would be in practice.

**Comment by max-sixty** [comment]
@andersy005 do you want to add to whatsnew?

Either way we can merge later

**Comment by andersy005** [comment]
Thank you to everyone who provided feedback/review....

FYI, the GitHub action failed to run after the merge: https://github.com/pydata/xarray/actions/runs/376810413

The logs show this error message:


> ...and actions/github-script@v3 are not allowed to be used in pydata/xarray. Actions in this workflow must be: within a repository owned by pydata.


It looks like the admins of the "pydata/xarray" repo have set the policy to be “Allow local actions only”. This option only allows the actions defined in the repositories within the same  organization “pydata” to be used. Can someone with admin privileges enable the “Allow all actions” option from the repo settings: https://github.com/pydata/xarray/settings/actions?

**Comment by jhamman** [comment]
@andersy005 - done! I manually triggered the CI and it seems to be working now: https://github.com/pydata/xarray/runs/1437420208?check_suite_focus=true
